src/ukify/ukify.py

```diff
@@ -572,21 +572,49 @@ def pe_add_sections(uki: UKI, output: str):
         else:
             new_section.IMAGE_SCN_CNT_INITIALIZED_DATA = True
 
-        pe.__data__ = pe.__data__[:] + bytes(new_section.PointerToRawData - len(pe.__data__)) + data + bytes(new_section.SizeOfRawData - len(data))
+        # Special case, mostly for .sbat: the stub will already have a .sbat section, but we want to append
+        # the one from the kernel to it. It should be small enough to fit in the existing section, so just
+        # swap the data.
+        for i, s in enumerate(pe.sections):
+            if s.Name.rstrip(b"\x00").decode() == section.name and new_section.Misc_VirtualSize <= s.SizeOfRawData:
```

Comment:
@medhefgo this is very hacky, and of course only works as long as we have enough space... As far as I can see with python3-pefile it is not possible to remove sections, or am I missing some api somewhere? Ideally we'd just read the sbat from the stub, then delete the section, and append a new one normally, but can't find an easy way to do that

Reply:
python-pefile's write support is basically just for changing header values as long as the file layout doesn't change (or there is enough space to fit a new header structure). Our code for adding sections is really just abusing private fields to do its job. So if you want to remove a section you'd have to do it manually.

Reply:
You may wanna make sure this works correctly for those bad old stubs (from centos?) that don't align/pad things correctly.

Reply:
> You may wanna make sure this works correctly for those bad old stubs (from centos?) that don't align/pad things correctly.

Aren't those fixed up earlier in the function?

Reply:
> So if you want to remove a section you'd have to do it manually.

Weeeell, 512 bytes of SBAT ought to be enough for everybody. Feels like I heard this somewhere before...

Reply:
> Aren't those fixed up earlier in the function?

Yeah, just saying you should double check. :P

Reply:
> Weeeell, 512 bytes of SBAT ought to be enough for everybody. Feels like I heard this somewhere before...

You should make it an error if an sbat section exists that isn't big enough then.

Reply:
> > Aren't those fixed up earlier in the function?
> 
> Yeah, just saying you should double check. :P

Do you know where can I get one of those? Old centos 7 packages?

Reply:
I believe it was ye olde centos 8 mkosi builds that needed this (as always).

Reply:
centos 8 builds and boots fine, ship it